Add interface to register notebooks

johnsca opened the pull request.

johnsca committed:
Add interface to register notebooks

kwmonroe commented on lib/charms/layer/apache_zeppelin.py:
```diff
@@ -168,3 +170,22 @@ def close_ports(self):
+    Helper for interacting with the Appache Zeppelin REST API.
+    """
+    def _url(self, *parts):
+        url = 'http://localhost:9090/api/'
```
Let's use the `DistConfig().port('zeppelin')` port instead of 9090:

https://github.com/juju-solutions/layer-apache-zeppelin/blob/master/dist.yaml#L32

pengale replied:
Agree. Let us not hard code any of the things.

Also: do we need to deal with the case where there are multiple zeppelins, and one is acting as a master?

johnsca replied:
Since these requests are coming in over the relation, they will apply to all units, which is the right thing to do.  (Every unit needs a copy of the notebook or interpreter settings.)

kwmonroe commented on lib/charms/layer/apache_zeppelin.py:
```diff
@@ -168,3 +170,22 @@ def close_ports(self):
+
+    def import_notebook(self, contents):
+        response = requests.post(self._url('notebook'), data=contents)
+        response.raise_for_status()
```
Feels like we should try/catch an error here.  Maybe pass on HTTP or ConnectionError in case they're transient and we want to try again, but fail on generic exception in case the api call just totes broke.

kwmonroe commented on lib/charms/layer/apache_zeppelin.py:
```diff
@@ -168,3 +170,22 @@ def close_ports(self):
+        return response.json()['body']
+
+    def delete_notebook(self, notebook_id):
+        requests.delete(self._url('notebook', notebook_id))
```
Feels like we need to catch something here too.

johnsca committed:
Refactored to acknowledge accept, reject, and remove

johnsca committed:
Fixed issues registering notebook

johnsca committed:
Fixed import API status code

johnsca committed:
Added xenial

pengale commented on lib/charms/layer/apache_zeppelin.py:
```diff
@@ -168,3 +170,23 @@ def close_ports(self):
+    def import_notebook(self, contents):
+        response = requests.post(self._url('notebook'), data=contents)
+        if response.status_code != 201:
+            return None
```
I would .raise_for_status() here. We want to be transparent about why things failed, if they failed.

johnsca replied:
The import specifically returns an accepted / rejected response to the requesting charm, so that the error can be surfaced closer to where the (presumable) cause is.  `raise_for_status` here would just cause the Zeppelin charm to go into an error state and become unusable.  We should avoid charm error states whenever possible, especially in the case where it may not be the charm's fault (e.g., if it gets passed an invalid notebook).

pengale replied:
That make sense. +1 on this bit of code.

pengale commented on lib/charms/layer/apache_zeppelin.py:
```diff
@@ -168,3 +170,23 @@ def close_ports(self):
+        return response.json()['body']
+
+    def delete_notebook(self, notebook_id):
+        requests.delete(self._url('notebook', notebook_id))
```
Would also like .raise_for_status() here.

pengale reviewed (commented):
You beat me to the punch on this one :-)

Overall, looks good, but I would like some more transparency on errors.

johnsca committed:
Added support for modifying interpreter settings via relation

johnsca committed:
Support Juju 2.0 resources

johnsca committed:
Avoid hard-coding port

johnsca committed:
Remove support for modifying interpreterGroup and options, as they do not affect any change

johnsca committed:
Fix Spark interpreter master property

johnsca committed:
Fix race condition in re/starting API

The API server can take a bit of time to come up after starting or
restarting.  This can cause other handlers to fail with
ConnectionRefused.

When the handler fails within the same hook context as the installation,
it can leave the charm in a state where the next hook will re-attempt
the installation, which isn't handled well.  We need to improve support
for install being called again.

johnsca committed:
Remove caveat about Spark interpreter master property from README

johnsca committed:
Fixed 2.0 resource deployment and enabled implementation class override

johnsca committed:
Changed implementation of restart

johnsca committed:
Clean up install and wait_for_stop

johnsca committed:
Fix handling of resource_get on Juju 1.25

kwmonroe commented on lib/charms/layer/apache_zeppelin.py:
```diff
@@ -168,3 +240,44 @@ def close_ports(self):
+    def delete_notebook(self, notebook_id):
+        requests.delete(self._url('notebook/', notebook_id))
+
+    def modify_interpreter(self, interpreter_name, properties):
```
Danger!  this requires zeppelin to be started.  In the reactive handlers, you are guarding calls to this function by `zeppelin.started`, but that state isn't always true.  For example, if you restart the unit post deployment, zeppelin will not auto-start, yet the `zeppelin.started` state will be set from the initial deployment.

We either need to jps test for zeppelin in this method (and not call requests.get if zepp is not started), or jps test from the reactive handler (starting zepp if necessary) before calling this method.

kwmonroe commented on reactive/zeppelin.py:
```diff
@@ -1,38 +1,53 @@
+def update_spark_master(spark):
+    zepp = Zeppelin.get()
+    api = ZeppelinAPI()
+    api.modify_interpreter('spark', properties={
```
Need to make sure zepp is actually running before calling this, else the `requests.get` to localhost:9090 will fail.

kwmonroe commented on reactive/zeppelin.py:
```diff
@@ -47,3 +62,58 @@ def report_blocked():
+    api = ZeppelinAPI()
+    for interpreter in client.interpreter_changes():
+        try:
+            api.modify_interpreter(interpreter['name'],
```
Same as above, need to make sure zepp is actually running before calling this, else the `requests.get` to localhost:9090 will fail.

kwmonroe reviewed (commented):
There's an edge case where calls to `modify_interpreter` will fail.  If the unit is rebooted post deployment, zeppelin will not auto start, yet the `zeppelin.started` state will still be set.

Calls to `request.get` will fail if zeppelin isn't actually running.

johnsca committed:
Clean up install and spark master logic

johnsca committed:
Work around charm store resource upload issue

johnsca committed:
Changed resource extension

johnsca committed:
Fix working with older versions of apache-spark

johnsca committed:
Fix falling back to S3

johnsca committed:
Another fix to fall-through resource logic

pengale commented on lib/charms/layer/apache_zeppelin.py:
```diff
@@ -13,15 +24,37 @@ class Zeppelin(object):
+        impl = layer.options('apache-zeppelin')['implementation_class']
+        module_name, cls_name = impl.rsplit('.', 1)
+        cls = getattr(import_module(module_name), cls_name)
+        return cls()
```
This doesn't feel Pythonic. I'd expect another layer to just subclass this class, and override what they want. Is there a reason for them not to do so?

johnsca replied:
I agree, it's not pretty.  The issue is that the reactive handlers in the base layer don't know anything about layers containing subclasses, so they can't instantiate the correct class.  This was my attempt to enable the base layer to be aware of the desired subclass, but I'm more than open to a better way to accomplish this.

pengale replied:
Hmmm ... I think that I'd rather see one of the following:

1) Override `__new__` with pretty much exactly this logic. That would allow people to just call Zeppelin(), without also calling something that looks like a Java getter :-)  You then have to explain the magic in a doc, but you have to explain it, anyway.

2) Do the logic you do here, but in the import line in the reactive handlers. If the implementation_class config isn't set, you just import Zeppelin. Otherwise, you import_module the thing in the implementation_class config.

3) Make people just write a custom Zeppelin layer for a custom Zeppelin subclass. I like this one the most, but I get why you didn't want to do it :-)

johnsca replied:
I went with the getter to make it clear that it's a factory method.  I kind of like the `__new__` idea, but it feels a bit subtle and prone to surprise.

Another option would be to make charms.layer.apache_zeppelin a submodule with an `__init__.py` and `implementation.py` or similar, and then the layer could override the import in `__init__.py` to set the subclass to use.

However, I think a best approach would be to do more refactoring to eliminate the need for the subclass entirely.  /me ponders.

pengale replied:
If you're going the factory approach, then I'd just rename .get to .factory (as done here http://python-3-patterns-idioms-test.readthedocs.io/en/latest/Factory.html). It does the exact same thing, but I think that it flags what you're doing more clearly.

johnsca committed:
Refactor master update to better support layers

ktsakalozos commented on lib/charms/layer/apache_zeppelin.py:
```diff
@@ -144,18 +204,50 @@ def start(self):
+        start = datetime.now()
+        while utils.jps("zeppelin"):
+            time.sleep(1)
+            if datetime.now() - start > timeout:
```
We get this here. We probably miss a .seconds like (datetime.now() - start).seconds

```
File "/var/lib/juju/agents/unit-apache-zeppelin-0/charm/hooks/spark-relation-departed", line 19, in <module>
2016-10-03 18:30:04 INFO spark-relation-departed     main()
2016-10-03 18:30:04 INFO spark-relation-departed   File "/var/lib/juju/agents/unit-apache-zeppelin-0/.venv/lib/python3.5/site-packages/charms/reactive/__init__.py", line 78, in main
2016-10-03 18:30:04 INFO spark-relation-departed     bus.dispatch()
2016-10-03 18:30:04 INFO spark-relation-departed   File "/var/lib/juju/agents/unit-apache-zeppelin-0/.venv/lib/python3.5/site-packages/charms/reactive/bus.py", line 434, in dispatch
2016-10-03 18:30:04 INFO spark-relation-departed     _invoke(other_handlers)
2016-10-03 18:30:04 INFO spark-relation-departed   File "/var/lib/juju/agents/unit-apache-zeppelin-0/.venv/lib/python3.5/site-packages/charms/reactive/bus.py", line 417, in _invoke
2016-10-03 18:30:04 INFO spark-relation-departed     handler.invoke()
2016-10-03 18:30:04 INFO spark-relation-departed   File "/var/lib/juju/agents/unit-apache-zeppelin-0/.venv/lib/python3.5/site-packages/charms/reactive/bus.py", line 291, in invoke
2016-10-03 18:30:04 INFO spark-relation-departed     self._action(*args)
2016-10-03 18:30:04 INFO spark-relation-departed   File "/var/lib/juju/agents/unit-apache-zeppelin-0/charm/reactive/zeppelin.py", line 51, in stop_zeppelin
2016-10-03 18:30:04 INFO spark-relation-departed     zepp.stop()
2016-10-03 18:30:04 INFO spark-relation-departed   File "lib/charms/layer/apache_zeppelin.py", line 255, in stop
2016-10-03 18:30:04 INFO spark-relation-departed     self.wait_for_stop(30)
2016-10-03 18:30:04 INFO spark-relation-departed   File "lib/charms/layer/apache_zeppelin.py", line 246, in wait_for_stop
2016-10-03 18:30:04 INFO spark-relation-departed     if datetime.now() - start > timeout:
2016-10-03 18:30:04 INFO spark-relation-departed TypeError: unorderable types: datetime.timedelta() > int()

```

pengale replied:
Another fix would just be to use time.time() for this. You're just measuring local system time, so you don't actually want the overhead of using datetime.

ktsakalozos-canonical committed:
Adding init scripts to Zeppelin

pengale commented on lib/charms/layer/apache_zeppelin.py:
```diff
@@ -115,6 +117,29 @@ def setup_zeppelin_config(self):
+            template_path = '/etc/init/zeppelin.conf'
+            template_name = 'upstart.conf'
+        if os.path.exists(template_path):
+            os.remove(template_path)
```
Would prefer to move this to a backup file instead. If we run setup_init_scripts twice, we still blow up the backup ... but at least it gives a sysadmin a chance to recover if the below goes poorly.

ktsakalozos replied:
The last commit should address this comment. We keep the last config with a .backup extension.

pengale commented on lib/charms/layer/apache_zeppelin.py:
```diff
@@ -231,15 +248,12 @@ def wait_for_stop(self, timeout):
         while utils.jps("zeppelin"):
             time.sleep(1)
-            if datetime.now() - start > timeout:
+            if (datetime.now() - start).seconds > timeout:
```
Using time.time() to set start and this value would be cleaner, but those code is functional, if a bit odd.

ktsakalozos replied:
The last commit should address this comment. Replaced datetime with time.

pengale reviewed (approved):
+1

ktsakalozos-canonical committed:
Keep init script backup. Use time instead of datetime.

johnsca committed:
Renamed factory method to be more clear

pengale commented:
LGTM/+1
